File: app/src/main/java/org/wikipedia/feed/topread/TopReadListCard.java
package org.wikipedia.feed.topread;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;

import org.wikipedia.R;
import org.wikipedia.dataclient.WikiSite;
import org.wikipedia.feed.model.CardType;
import org.wikipedia.feed.model.ListCard;
import org.wikipedia.util.DateUtil;
import org.wikipedia.util.L10nUtil;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

public class TopReadListCard extends ListCard<TopReadItemCard> {
    @NonNull private final TopRead articles;

    public TopReadListCard(@NonNull TopRead articles, @NonNull WikiSite wiki) {
        super(toItems(articles.articles(), wiki), wiki);
        this.articles = articles;
    }

    @NonNull
    @Override public String title() {
        return L10nUtil.getStringForArticleLanguage(wikiSite().languageCode(), R.string.view_top_read_card_title);
    }

    @Nullable
    @Override public String subtitle() {
        return DateUtil.getFeedCardDateString(articles.date());
    }

    @NonNull
    @Override public CardType type() {
        return CardType.TOP_READ_LIST;
    }

    @NonNull
    public String footerActionText() {
        return L10nUtil.getStringForArticleLanguage(wikiSite().languageCode(), R.string.view_top_read_card_action);
    }

    @NonNull
    @VisibleForTesting
    public static List<TopReadItemCard> toItems(@NonNull List<TopReadArticles> articles,
                                                @NonNull WikiSite wiki) {
        List<TopReadItemCard> cards = new ArrayList<>();
        for (TopReadArticles article : articles) {
            cards.add(new TopReadItemCard(article, wiki));
        }
        return cards;
    }

    @Override
    protected int dismissHashCode() {
        return (int) TimeUnit.MILLISECONDS.toDays(articles.date().getTime()) + wikiSite().hashCode();
    }
}

File: app/src/main/java/org/wikipedia/feed/topread/TopReadListCard.kt
package org.wikipedia.feed.topread

import android.os.Parcelable
import kotlinx.parcelize.Parcelize
import org.wikipedia.R
import org.wikipedia.dataclient.WikiSite
import org.wikipedia.dataclient.page.PageSummary
import org.wikipedia.feed.model.CardType
import org.wikipedia.feed.model.ListCard
import org.wikipedia.util.DateUtil
import org.wikipedia.util.L10nUtil

@Parcelize
class TopReadListCard(private val articles: TopRead, val site: WikiSite) :
    ListCard<TopReadItemCard>(toItems(articles.articles, site), site), Parcelable {

    override fun title(): String {
        return L10nUtil.getStringForArticleLanguage(wikiSite().languageCode, R.string.view_top_read_card_title)
    }

    override fun subtitle(): String {
        return DateUtil.getShortDateString(articles.localDate)
    }

    override fun type(): CardType {
        return CardType.TOP_READ_LIST
    }

    fun footerActionText(): String {
        return L10nUtil.getStringForArticleLanguage(wikiSite().languageCode, R.string.view_top_read_card_action)
    }

    override fun dismissHashCode(): Int {
        return articles.localDate.toEpochDay().toInt() + wikiSite().hashCode()
    }

    companion object {
        fun toItems(articles: List<PageSummary>, wiki: WikiSite): List<TopReadItemCard> {
            return articles.map { TopReadItemCard(it, wiki) }
        }
    }
}
